### ISSUE TITLE
General question on SHAP interaction values: main effect and interaction effect

### ISSUE THREAD

**Issue by liuyanguu** [shap/shap#707]
Hey @slundberg, Thanks again for this great package! According to your paper [Consistent Individualized Feature Attribution for Tree
Ensembles](https://arxiv.org/abs/1802.03888), it says "SHAP **interaction values** can be interpreted as the difference between the A. SHAP values for feature i when feature j is **present** and B. the SHAP values for feature i when feature j is **absent**."
So literally, I suppose the interaction effect could be calculated by 
```Interaction effect = SHAP value for i in model A (full model) - SHAP values for i in model B (w/o j)``` ? Is it appropriate to calculate in this way?

Then in Figure 10, it says ```interaction effect + main effect = SHAP value for i in model A``` (the dependence plot)

Then does it mean the main effect is just the _SHAP values for i in model B_, where feature j is absent?

I somehow got myself confused. Any advice is highly appreciated.
Lots of thanks,
Yang Liu

**Comment by HexFluid** [comment]
Hi Yang,

I happened to think the same and did a quick test on my own dataset. I trained two RF models: one with full input features, and the other with the feature j absent. Then I calculated the shap values of both models. I compared the shap interaction value phi_{i,j} calculated from the full-feature RF and the shap value difference (phi_i^{full feature} - phi_i^{j absent})/2, but the results are very different. I think the reason is that, when training the RF model to map all the input features except j to the outputs, the attributes originally from feature j are distributed to other features in an unclear way. Thus, when explaining this trained RF using shap values, it does not make much sense. (i.e., the idea of using phi_i^{j absent} is doubtful)

An alternative approach is to calculate the shap value difference (phi_i^{unmute} - phi_i^{mute j})/2 while mute/unmute feature j. I trained one RF model to map the full input features to the outputs. Based on this, I calculated two shap values: the first uses the original input data, and the second uses the input data where the feature j is manually overwritten as its mean value (all data samples have the same value of j, and thus the effect of j is muted). The results show reasonably good agreement with the shap interaction value. I think the minor difference comes from the treatment of feature j: here we just used its mean, but the shap interaction calculation uses a more rigorously probability-based method to estimate j.

The results of the above two methods are shown in the following fig.

![shap_inter_test](https://user-images.githubusercontent.com/46558908/130353991-a8bc7302-69f5-48ed-ab40-a2b09b08b7b7.png)

Hope it helps.

Xiao
